keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,897 @@
+
+Same as the current [`controller.kubernetes.io/pod-deletion-cost`](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost) annotation.
+
+__USAGE PATTERN 1 - the annotation kept up-to-date with a controller:__
+
+- PROBLEM 1: every update requires a PATCH call to the kube-apiserver (probably overloading it)
+- PROBLEM 2: this is wasteful for deployments that rarely scale down
+   - calculating and updating the cost could be expensive, and that work is wasted if not actually used to downscale
+
+__USAGE PATTERN 2 - the annotation is updated only when downscaling:__
+
+- PROBLEM 1: existing scaling tools like HorozontalPodAutoscaler can't be used
+   - the annotation/status would need to be updated BEFORE downscaling, and we can't predict when HorozontalPodAutoscaler will downscale
+   - therefore, users must write their own scalers that update the annotation/status before downscaling (making it inaccessible for most users)
+- PROBLEM 2: after each downscale, any annotations added by the scaler will need to be cleared (they will become out of date)
+   - related to this, we must wait to clear them before we can start the next downscale (or the old annotations will impact the new scale)
+- PROBLEM 3: costs may be out-of-date by the time the controller picks which pod to downscale
+   - some apps like airflow/spark may have quite rapidly changing "best" pods to kill
+
+### ALTERNATIVE 2: Pod-Deletion-Cost http/exec probe
+
+A new http/exec probe is created for Pods which returns their current pod-deletion-cost.
+
+__USAGE PATTERN 1 - probes are used every X seconds to update a Pod status field:__
+
+- (Suffers from the same problems as "USAGE PATTERN 1" of "ALTERNATIVE 1")
+
+__USAGE PATTERN 2 - probes are ONLY used when downscaling:__
+
+- PROBLEM 1: the controller cannot make a probe request to Pods (this must be done by the Node's kubelet)
+  - to solve this you would need a complex system that has the controller "mark" the pods for probing by the kubelet
+- PROBLEM 2: to find the "best" Pod to kill, we would need to probe every single Pod, this is not scalable
+  - to solve this, you would need to use a heuristic approach, e.g. only checking a sample of Pods and returning the lowest cost from the sample
+
+### ALTERNATIVE 3: Pod-Deletion-Cost API
+
+A central user-managed API is queried by the controller and returns the pod-deletion-cost of each Pod.
+(Rather than returning a list of `chosen_pods` and `tied_pods` like in this KEP)
+
+- PROBLEM 1: the user-managed API can't exit-early
+  - because a pod-deletion-cost must be returned for each Pod, the API can't have the concept of a "free" Pod, which if found can be immediately returned without checking the other Pods
+- PROBLEM 2: the user-managed API may calculate the pod-deletion-cost of more Pods than necessary
+  - the API is unaware of how many Pods are actually planned to be removed, so will calculate the pod-deletion-cost of more Pods than is necessary
```

Comment:
You could have the API only have the cost of the lowest cost pods.

If the replicaset controller needs to downscale more than that number of pods, it first removes the ones for which it knows the cost, then waits for the API to update with the cost of the next lowest cost pods.

Reply:
@alculquicondor I don't think there is a need to return the "cost" value, the Pod-Picker API can just "pick" the ones it thinks are best to remove, and return "undecided" on the ones it can't decide between. 

In the end, the RepliaSet controller will just be "choosing" from the lowest cost ones anyway, so removing the cost just simplifies the API contract.

Reply:
@alculquicondor an ordered list was actually my original design, but I realized that doesn't handle situations when there are ties (but NOT all are tied).

For example, if the Pod-Picker was internally assigning the following "deletion costs" to each pod:
- `pod_1` = `cost 1`
- `pod_2` = `cost 5`
- `pod_3` = `cost 5`
- `pod_4` = `cost 5`
- `pod_5` = `cost 10`

When `number_of_pods_requested = 2`, there is no way for the Pod-Picker to indicate that `pod_1` should definitely be killed, as it will have to return `[pod_1, pod_2, pod_3, pod_4]` (or just randomly pick between `[pod_2, pod_3, pod_4]`).  

Ideally, the ReplicaSet controller should know to kill `pod_1`, and then continue with the [existing down-scale selection criteria](https://github.com/kubernetes/kubernetes/blob/dd3a30fdbbacbca1d6053efeb4bb474564caf7fe/pkg/controller/controller_utils.go#L844-L880) to decide which of `[pod_2, pod_3, pod_4]` is going to be killed:
1. "Pods with less time having Ready condition"
2. "Pods with higher restart count"
3. "Pods with newer creation time"

NOTE: with the Pod-Picker API proposed in this KEP, we would have the following inputs/outputs:
- INPUT:
   - `number_of_pods_requested` = `2`
   - `candidate_pods` = `[pod_1, pod_2, pod_3, pod_4, ..., pod_n]`
- REUTRN:
   - `chosen_pods` = `[pod_1]`
   - `tied_pods` = `[pod_2, pod_3, pod_4]`

Reply:
Then include the costs and let the ReplicaSet controller do the next comparisons :)

Reply:
@alculquicondor is it not simpler to just return two lists? 

What the ReplicaSet controller is actually doing is asking for the Pod-Picker API to nominate pods for deletion (not to rank them), so adding an arbitrary "cost" just hides the intent of the API and introduces unnecessary complexity in both producing and interpreting the API response.

Also, not all Pod-Picker use-cases have an internal concept of "cost", for example, a Pod-Picker may be ensuring the number of shards in each region stays above a threshold, which is not really a "deletion cost" in the traditional sense.

Reply:
Having the cost for each pod could have the advantage that the information is useful for a few scale downs. Or, let's say HPA decides that it needs to reduce 10 replicas. While the pod picker is doing the work, HPA decides that it only needs to reduce 5 replicas. Instead of waiting for the pod picker again, the replicaset controller could use the information that is already calculated.

Both approaches have advantages and disadvantages. I think the main question is whether a resource like this would work, as an alternative to having a webhook.

Reply:
I've been trying to boost a 3rd approach, a query API using basic metrics (Metrics Server sort of thing) but involving the API server and its mechanisms for authentication, rate limiting, etc. You'd use APIService to configure the extension API, and if you don't, HPA can fall back to existing behavior.

Some of the flavors of this would be agnostic about whether you use a CustomResourceDefinition or an APIService; some flavours only really work with an APIService - the same way that we don't use CRDs for metric scrapes.

Reply:
> Having the cost for each pod could have the advantage that the information is useful for a few scale downs.
> 
> Scenario: Let's say HPA decides that it needs to reduce 10 replicas. While the pod picker is doing the work, HPA decides that it only needs to reduce 5 replicas. Instead of waiting for the pod picker again, the replicaset controller could use the information that is already calculated.

@alculquicondor I believe once the ReplicaSet has decided to kill a certain number of Pods, it will always go ahead and do that (see the code for [`ReplicaSetController.manageReplicas()`](https://github.com/kubernetes/kubernetes/blob/876d4e0ab029ac7c314bb0e033bdd036531fe426/pkg/controller/replicaset/replica_set.go#L543)). That is, once an HPA has made a scaling decision (and lowered the `replicas` count), the ReplicaSet controller will have already initiated a scale-down, regardless of if the HPA makes another scaling decision while the previous scale-down was being processed.

> Both approaches have advantages and disadvantages. I think the main question is whether a resource like this would work, as an alternative to having a webhook.

I am not sure I understand, are you proposing we implement the Pod-Picker API as a new kind of Webhook?
Note that all the current webhooks are focused on admission control and live in `apiVersion: admissionregistration.k8s.io/v1`.

Reply:
@sftim I am not 100% sure what you are proposing, but see my response up in  https://github.com/kubernetes/enhancements/pull/3190#discussion_r830778891